### PR TITLE
[SPARK-25357][SQL] Add metadata to SparkPlanInfo to dump more information like file path to event log

### PR DESCRIPTION
**Pull request by LantaoJin**
## What changes were proposed in this pull request?

Field metadata removed from SparkPlanInfo in #18600 . Corresponding, many meta data was also removed from event SparkListenerSQLExecutionStart in Spark event log. If we want to analyze event log to get all input paths, we couldn't get them. Instead, simpleString of SparkPlanInfo JSON only display 100 characters, it won't help.

Before 2.3, the fragment of SparkListenerSQLExecutionStart in event log looks like below (It contains the metadata field which has the intact information):
>{"Event":"org.apache.spark.sql.execution.ui.SparkListenerSQLExecutionStart", Location: InMemoryFileIndex[hdfs://cluster1/sys/edw/test1/test2/test3/test4..., "metadata": {"Location": "InMemoryFileIndex[hdfs://cluster1/sys/edw/test1/test2/test3/test4/test5/snapshot/dt=20180904]","ReadSchema":"struct<snpsht_start_dt:date,snpsht_end_dt:date,am_ntlogin_name:string,am_first_name:string,am_last_name:string,isg_name:string,CRE_DATE:date,CRE_USER:string,UPD_DATE:timestamp,UPD_USER:string>"}

After #18600, metadata field was removed.
>{"Event":"org.apache.spark.sql.execution.ui.SparkListenerSQLExecutionStart", Location: InMemoryFileIndex[hdfs://cluster1/sys/edw/test1/test2/test3/test4...,

So I add this field back to SparkPlanInfo class. Then it will log out the meta data to event log. Intact information in event log is very useful for offline job analysis.

## How was this patch tested?
Unit test

### CODE REVIEW

**Comment by LantaoJin** [comment]
@cloud-fan @gatorsmile @dongjoon-hyun , kindly help to review.

**Comment by LantaoJin** [comment]
@wangyum  @hvanhovell

**Comment by LantaoJin** [comment]
A scenario here is after an application completed, there is no way to know the intact file path of File Scan Exec if the path width is longer than 100 chars.

File: sql/core/src/main/scala/org/apache/spark/sql/execution/DataSourceScanExec.scala [hunk 1]
```diff
@@ -54,7 +54,7 @@ trait DataSourceScanExec extends LeafExecNode with CodegenSupport {
   override def simpleString: String = {
     val metadataEntries = metadata.toSeq.sorted.map {
       case (key, value) =>
-        key + ": " + StringUtils.abbreviate(redact(value), 100)
```
[the remark this answers is not in the document]
Comment (LantaoJin) [reply sql/core/src/main/scala/org/apache/spark/sql/execution/DataSourceScanExec.scala]: I think it’s overkill to parameterizing it. And Spark user doesn’t care about it, no one will reset it before submitting app. Besides, simply raise up to 1000 also can resolve the problem on most cases, but longer than 1000 chars is still meanlessness.

[the remark this answers is not in the document]
Comment (maropu) [reply sql/core/src/main/scala/org/apache/spark/sql/execution/DataSourceScanExec.scala]: We already have `spark.debug.maxToStringFields` for debugging, so is it bad to add `spark.debug.XXXX`? Anyway, if most users don't care about this, I think changing the default number goes too far.

[the remark this answers is not in the document]
Comment (dongjoon-hyun) [reply sql/core/src/main/scala/org/apache/spark/sql/execution/DataSourceScanExec.scala]: This seems to cause a regression on Spark Web UI. Could you check that, @LantaoJin ?

In fact, the abbreviation is introduced over two years ago at Spark 2.0 intentionally for UI via [[SPARK-14476][SQL] Improve the physical plan visualization by adding meta info like table name and file path for data source](https://github.com/apache/spark/pull/12947). At least, we had better update the information of PR and JIRA.

**Review by dongjoon-hyun — CHANGES_REQUESTED** [review]
This may cause a regression on Spark UI, @LantaoJin . Could you check that?

In fact, this abbreviation is introduced over two years ago at Spark 2.0 via [[SPARK-14476][SQL] Improve the physical plan visualization by adding meta info like table name and file path for data source](https://github.com/apache/spark/pull/12947) . We had better update the PR and JIRA description.

**Comment by dongjoon-hyun** [comment]
@LantaoJin . Please check the following example in Spark UI; the hover text on `Scan parquet`.
```scala
scala> spark.range(20000).repartition(10000).write.mode("overwrite").parquet("/tmp/10000")
scala> spark.read.parquet("/tmp/10000/*").count
```

**Comment by LantaoJin** [comment]
Thanks @dongjoon-hyun . That would be a problem. Seems setting to 200 or 500 are cause a limited regression on hover text.

Hard code to 500 shows:
screen shot 2018-09-10 at 5 03 53 pm

**Comment by LantaoJin** [comment]
The purpose is logging meta info like file input file path to event log. So I revert the changes about simpleString and add back the metadata to SparkPlanInfo interface. This change will log metadata into event log in SparkListenerSQLExecutionStart and it is still benefited in #18600 .

File: sql/core/src/main/scala/org/apache/spark/sql/execution/SparkPlan.scala [hunk 2]
```diff
@@ -73,6 +73,11 @@ abstract class SparkPlan extends QueryPlan[SparkPlan] with Logging with Serializ
     super.makeCopy(newArgs)
   }
 
+  /**
+   * @return Metadata that describes more details of this SparkPlan.
+   */
+  def metadata: Map[String, String] = Map.empty
+
```
Comment (dongjoon-hyun) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/SparkPlan.scala]: cc @cloud-fan

**Comment by cloud-fan** [comment]
Although event log is in JSON format, it's mostly for internal usage, to be load by history server and used to build the Spark UI. For compatibility, we only focus on making history to be able to load event logs from different spark versions, not the event log itself. At the end it's still a log.

Metadata is a hack which I really hate to add back. Can you describe more details about your use case? Let's see if we can solve it with the Spark driver log.

**Comment by LantaoJin** [comment]
> Although event log is in JSON format, it's mostly for internal usage, to be load by history server and used to build the Spark UI.
AFAIK, there are more and more projects replay event log to analysis jobs offline, especially in a platform/infra team in a big company. Dr-elephant doesn't read event log, instead, query SHS to get information causing many problems like compatibility or data accuracy. In eBay we are building a system similar with Dr-elephant but much powerful. One of use cases in this system is building a data lineage and monitor the input/output path and data size for each application. Difference with Apache Altas who need attach a spark listener into the spark runtime, we choose to replay event log to build all context we need. Before 2.3, we can get above information from the `metadata` field in SQLExecutionStart event. Now it was removed. So I hope this PR could add it back. What's more is  make more probability on event log instead of only using in SHS.

**Comment by LantaoJin** [comment]
Spark driver log is always distributed on various client nodes and depends on the log4j configs. In a big company, it's hard to collect them all and I think it's better to used for debug not analyze.

**Comment by LantaoJin** [comment]
Thank you @cloud-fan for your reminding. We’ve handled the drop message case. Agree, I will update a commit tomorrow.

**Comment by cloud-fan** [comment]
ok to test

File: sql/core/src/main/scala/org/apache/spark/sql/execution/SparkPlanInfo.scala [hunk 3]
```diff
@@ -59,6 +57,12 @@ private[execution] object SparkPlanInfo {
       new SQLMetricInfo(metric.name.getOrElse(key), metric.id, metric.metricType)
     }
 
-    new SparkPlanInfo(plan.nodeName, plan.simpleString, children.map(fromSparkPlan), metrics)
+    // dump the file scan metadata (e.g file path) to event log
```
Comment (dongjoon-hyun) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/SparkPlanInfo.scala]: As a next step of reviews, did you have a chance to test this on your real environment at least TPCDS 1TB?
This seems to increase the event log traffic dramatically in the worst case. Can we have some comparison before and after this PR? @LantaoJin .

**Comment by cloud-fan** [comment]
thanks, merging to master/2.4/2.3!

**Pull request closed without merging**